### PR TITLE
Migrate from the php site

### PR DESCRIPTION
**Pull request by MilanRusev**
Currently without the scripts and fetching eth price

### CODE REVIEW

File: lib/sanbase/model/latest_btc_wallet_data.ex [hunk 1]
```diff
@@ -0,0 +1,20 @@
+defmodule Sanbase.Model.LatestBtcWalletData do
+  use Ecto.Schema
+  import Ecto.Changeset
+  alias Sanbase.Model.LatestBtcWalletData
+
+
+  @primary_key{:address, :string, []}
+  schema "latest_btc_wallet_data" do
+    # field :address, :string
+    field :satoshi_balance, :float
```
Comment (valo) [inline lib/sanbase/model/latest_btc_wallet_data.ex]: Are you sure this should be a `float` here? I think satoshis are not divisible: https://en.bitcoin.it/wiki/Satoshi_(unit)

File: lib/sanbase/model/latest_eth_wallet_data.ex [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+
+  @primary_key{:address, :string, []}
+  schema "latest_eth_wallet_data" do
+    # field :address, :string
+    field :balance, :float
+    field :last_incoming, :naive_datetime
+    field :last_outgoing, :naive_datetime
+    field :tx_in, :float
+    field :tx_out, :float
```
Comment (valo) [inline lib/sanbase/model/latest_eth_wallet_data.ex]: Should we use `decimal` for these values? I am a bit hesitant to use `float` to represent money amounts...

File: lib/sanbase_web/controllers/cashflow_controller.ex [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  def index(conn, _params) do
+    data = Cashflow.get_project_data
+
+    data = Enum.group_by(data,
+      fn(el) ->
+        %{project: el.project, coinmarketcap: el.coinmarketcap}
+      end,
+      fn(el) ->
+        el.wallet_data
+      end)
```
Comment (valo) [inline lib/sanbase_web/controllers/cashflow_controller.ex]: There is a shorthand for this:

```elixir
data = Cashflow.get_project_data
|> Enum.group_by(&Map.take(&1, [:project, :coinmarketcap]), &(&1.wallet_data))
```

I see there are other transformations on this data alter, which can be extracted into separate functions and added to the pipeline. This should make this code more readable.

File: lib/sanbase_web/controllers/cashflow_controller.ex [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    data = Enum.map(data,
+      fn({k, v}) ->
+        market_cap_usd = if (k.coinmarketcap !== nil), do: k.coinmarketcap.market_cap_usd, else: nil
+        wallet_data = Enum.filter(v, fn(x) -> x !== nil end)
```
Comment (valo) [inline lib/sanbase_web/controllers/cashflow_controller.ex]: Are you sure there could be `nil` values in the `v` here?

File: lib/sanbase_web/controllers/cashflow_controller.ex [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        balance = Enum.reduce(wallet_data, 0, fn(x, acc) -> x.balance + acc end)
+        %{
+          name: k.project.name,
+          ticker: k.project.ticker,
+          logo_url: k.project.logo_url,
+          market_cap_usd: market_cap_usd,
+          balance: balance,
+          wallets: wallet_data
+        }
+      end)
```
Comment (valo) [inline lib/sanbase_web/controllers/cashflow_controller.ex]: I would definitely try to split into a separate function. I think the `SanbaseWeb.CashflowView` makes sense to be used here, as this data is formatted for the presentation layer here. The upside of moving it to a separate function in the view is that we will be able to write unit tests for this function without touching the DB.

File: app/pages/cashflow/index.js [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    <script dangerouslySetInnerHTML={{ __html: `
+      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
+      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
+      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
+      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
+
+      ga('create', 'UA-100571693-1', 'auto');
+      ga('send', 'pageview');
+     `}} />
+  </Head>
```
Comment (valo) [inline app/pages/cashflow/index.js]: We should be able to extract these headers into separate components, but this is something that can be done later. For now I think we can merge this as it is and then create separate PRs to split this UI into smaller pieces that can be reused across the views.

File: app/pages/cashflow/index.js [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+                          )
+                        })}
+                          </tbody>
+                      </table>
+                  </div>
+              </div>
+          </div>
+      </div>
+  </div>
+  <script dangerouslySetInnerHTML={{ __html: `
```
Comment (valo) [inline app/pages/cashflow/index.js]: We should remove any uses of `dangerouslySetInnerHTML`, but that can be done in a subsequent PRs.

File: lib/sanbase/model/latest_btc_wallet_data.ex (hunk 1 above)
Comment (MilanRusev) [reply lib/sanbase/model/latest_btc_wallet_data.ex]: I agree, we may need to change the database types. Currently I made the same DB schema as the one for project transparency. We should refactor the DB datatypes for both projects.

File: lib/sanbase/model/latest_eth_wallet_data.ex (hunk 2 above)
Comment (MilanRusev) [reply lib/sanbase/model/latest_eth_wallet_data.ex]: The same thing for the DB schema being the the same as project transparency

File: lib/sanbase_web/controllers/cashflow_controller.ex (hunk 4 above)
Comment (MilanRusev) [reply lib/sanbase_web/controllers/cashflow_controller.ex]: Yes, if there is no wallet data for a project. The sql query is with a left join.

File: lib/sanbase/model/latest_btc_wallet_data.ex (hunk 1 above)
Comment (valo) [reply lib/sanbase/model/latest_btc_wallet_data.ex]: In that case we should probably keep it as it is and make subsequent PRs, which are fixing the schema

**Comment by valo** [comment]
I have some more comments on this, but let's merge it as it is and fix them later.

**Pull request merged**